service/worker/scheduler/workflow.go

```diff
@@ -626,23 +643,20 @@ func (s *scheduler) processTimeRange(
 	}
 
 	for {
-		var next getNextTimeResult
-		if !s.hasMinVersion(BatchAndCacheTimeQueries) {
-			// Run this logic in a SideEffect so that we can fix bugs there without breaking
-			// existing schedule workflows.
-			panicIfErr(workflow.SideEffect(s.ctx, func(ctx workflow.Context) interface{} {
-				return s.cspec.getNextTime(s.jitterSeed(), t1)
-			}).Get(&next))
-		} else {
-			next = s.getNextTime(t1)
-		}
+		next := s.getNextTime(t1)
```

Comment:
I think you can pull `next := s.getNextTime(t1)` into `for` init part, and then `next = s.getNextTime(next.Next)` into `for` step part. Then you could just inline t1 in this loop.

Reply:
I tried this and the code looks like:
```
    var next getNextTimeResult
    for next = s.getNextTime(t1); !(next.Next.IsZero() || next.Next.After(t2)); next = s.getNextTime(start) {
        // ... rest of loop, using next.Next instead of t1
    }
    return next.Next
```
It has to be declared outside of the loop so the return can use it. I'm not sure which version I like better but I guess I'll keep it.

Reply:
For me, this loop is very difficult to understand. Too many things are happening at the same time, and I have a very limited working memory. I would love to start to refactor this loop into something more manageable for me, if you do not mind.

Reply:
I think it's better now. There are four things:
1. `!s.hasMinVersion(BatchAndCacheTimeQueries) && !s.canTakeScheduledAction(manual, false)` can basically be ignored since it's for older versions.
2. `!manual && s.Info.UpdateTime.AsTime().After(next.Next)` this is the new clause, explained in the comment.
3. `!manual && end.Sub(next.Next) > catchupWindow` this is the catchup window, which is pretty simple.
4. `(*limit)--; *limit <= 0` there's a limit for backfills.
Besides those it just adds the start to the buffer.

We can try to refactor, but keep in mind the behavior, including side effects like markers, has to be exactly identical (unless using versioning calls), and if it's hard to verify that then we shouldn't change it.